File: packages/go/internal/rawhttp/rawhttp_test.go
package rawhttp

import (
	"bufio"
	"context"
	"net"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
)

// rawServer accepts one connection, captures the raw request bytes up to the
// end of headers (plus any body by content-length when parseable), and writes
// a fixed response.
func rawServer(t *testing.T, response string) (addr string, got chan string) {
	t.Helper()
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { l.Close() })
	got = make(chan string, 1)
	go func() {
		conn, err := l.Accept()
		if err != nil {
			return
		}
		defer conn.Close()
		r := bufio.NewReader(conn)
		var b strings.Builder
		for {
			line, err := r.ReadString('\n')
			if err != nil {
				break
			}
			b.WriteString(line)
			if line == "\r\n" {
				break
			}
		}
		got <- b.String()
		conn.Write([]byte(response))
	}()
	return l.Addr().String(), got
}

const okResponse = "HTTP/1.1 200 OK\r\nx-amz-request-id: R1\r\nContent-Length: 2\r\nConnection: close\r\n\r\nok"

func TestUnsignedLiteralBytes(t *testing.T) {
	addr, got := rawServer(t, okResponse)
	res, err := Do(context.Background(), "http://"+addr, &Request{
		Method: "PUT",
		Path:   "/bucket/k",
		Query:  map[string][]string{"partNumber": {"1"}},
		Headers: []Header{
			{"content-length", "-1"},
			{"x-weird", ""},
		},
		Body: []byte("abc"),
		Sign: false,
	})
	if err != nil {
		t.Fatal(err)
	}
	if res.Status != 200 || res.Header.Get("x-amz-request-id") != "R1" || string(res.Body) != "ok" {
		t.Errorf("response: %+v body %q", res, res.Body)
	}
	raw := <-got
	if !strings.HasPrefix(raw, "PUT /bucket/k?partNumber=1 HTTP/1.1\r\n") {
		t.Errorf("request line wrong:\n%s", raw)
	}
	if !strings.Contains(raw, "content-length: -1\r\n") {
		t.Errorf("content-length override not sent literally:\n%s", raw)
	}
	if !strings.Contains(raw, "x-weird: \r\n") {
		t.Errorf("empty header value not sent:\n%s", raw)
	}
	if strings.Contains(strings.ToLower(raw), "authorization") || strings.Contains(strings.ToLower(raw), "x-amz-date") {
		t.Errorf("unsigned request must carry no auth headers:\n%s", raw)
	}
}

func TestSignedRequest(t *testing.T) {
	addr, got := rawServer(t, okResponse)
	res, err := Do(context.Background(), "http://"+addr, &Request{
		Method:  "PUT",
		Path:    "/bucket/key",
		Headers: []Header{{"content-md5", "rL0Y20xC+Fzt72VPzMSk2A=="}},
		Body:    []byte("da"),
		Sign:    true,
		Creds:   aws.Credentials{AccessKeyID: "AK", SecretAccessKey: "SK"},
		Region:  "us-east-1",
	})
	if err != nil {
		t.Fatal(err)
	}
	if res.Status != 200 {
		t.Errorf("status %d", res.Status)
	}
	raw := <-got
	if !strings.Contains(raw, "Authorization: AWS4-HMAC-SHA256 Credential=AK/") {
		t.Errorf("missing/typo Authorization:\n%s", raw)
	}
	// The signature must cover the step's content-md5 header but never
	// content-length (which the corpus overrides on signed requests).
	lower := strings.ToLower(raw)
	authLine := lower[strings.Index(lower, "authorization:"):]
	authLine = authLine[:strings.Index(authLine, "\r\n")]
	if !strings.Contains(authLine, "content-md5") {
		t.Errorf("content-md5 not in SignedHeaders: %s", authLine)
	}
	if strings.Contains(authLine, "content-length") {
		t.Errorf("content-length must not be signed: %s", authLine)
	}
	if !strings.Contains(raw, "X-Amz-Content-Sha256: ") || !strings.Contains(raw, "X-Amz-Date: ") {
		t.Errorf("missing sigv4 headers:\n%s", raw)
	}
	if !strings.Contains(raw, "Content-Length: 2\r\n") {
		t.Errorf("default content-length missing:\n%s", raw)
	}
}

func TestAuthorizationOverrideWins(t *testing.T) {
	addr, got := rawServer(t, okResponse)
	_, err := Do(context.Background(), "http://"+addr, &Request{
		Method:  "GET",
		Path:    "/bucket",
		Headers: []Header{{"authorization", ""}},
		Sign:    true,
		Creds:   aws.Credentials{AccessKeyID: "AK", SecretAccessKey: "SK"},
		Region:  "us-east-1",
	})
	if err != nil {
		t.Fatal(err)
	}
	raw := <-got
	if !strings.Contains(raw, "authorization: \r\n") {
		t.Errorf("authorization override must win over the signed value:\n%s", raw)
	}
	if strings.Contains(raw, "AWS4-HMAC-SHA256") {
		t.Errorf("signed authorization leaked despite override:\n%s", raw)
	}
}

func TestParseXMLError(t *testing.T) {
	code, msg := ParseXMLError([]byte(`<?xml version="1.0"?><Error><Code>NoSuchKey</Code><Message>nope</Message></Error>`))
	if code != "NoSuchKey" || msg != "nope" {
		t.Errorf("got %q %q", code, msg)
	}
	if code, _ := ParseXMLError([]byte("not xml")); code != "" {
		t.Errorf("non-xml should give empty code, got %q", code)
	}
	if code, _ := ParseXMLError(nil); code != "" {
		t.Errorf("empty body should give empty code, got %q", code)
	}
}

File: packages/js/test/rawhttp.test.js
import test from 'node:test'
import assert from 'node:assert/strict'
import net from 'node:net'
import { once } from 'node:events'
import { rawRequest, parseResponse, parseXmlError } from '../lib/rawhttp.js'

const OK_RESPONSE = 'HTTP/1.1 200 OK\r\nx-amz-request-id: R1\r\nContent-Length: 2\r\nConnection: close\r\n\r\nok'

/** Raw TCP server: captures the request head, replies with `response`. */
async function rawServer (response) {
  let captured = ''
  const server = net.createServer((socket) => {
    socket.on('data', (chunk) => {
      captured += chunk.toString('latin1')
      if (captured.includes('\r\n\r\n')) {
        socket.end(response)
      }
    })
  })
  server.listen(0, '127.0.0.1')
  await once(server, 'listening')
  return {
    url: `http://127.0.0.1:${server.address().port}`,
    head: () => captured,
    close: () => new Promise((r) => server.close(r))
  }
}

const creds = { accessKeyId: 'AK', secretAccessKey: 'SK' }

test('unsigned requests send literal bytes with no auth headers', async () => {
  const srv = await rawServer(OK_RESPONSE)
  try {
    const res = await rawRequest(srv.url, {
      method: 'PUT',
      path: '/bucket/k',
      query: { partNumber: ['1'] },
      headers: [['content-length', '-1'], ['x-weird', '']],
      body: new TextEncoder().encode('abc'),
      sign: false
    })
    assert.equal(res.status, 200)
    assert.equal(res.headers['x-amz-request-id'], 'R1')
    assert.equal(Buffer.from(res.body).toString(), 'ok')
    const head = srv.head()
    assert.ok(head.startsWith('PUT /bucket/k?partNumber=1 HTTP/1.1\r\n'), head)
    assert.ok(head.includes('content-length: -1\r\n'), 'content-length override sent literally')
    assert.ok(head.includes('x-weird: \r\n'), 'empty header value sent')
    assert.ok(!/authorization|x-amz-date/i.test(head), 'no auth headers when unsigned')
  } finally {
    await srv.close()
  }
})

test('signed requests cover step headers but never content-length', async () => {
  const srv = await rawServer(OK_RESPONSE)
  try {
    await rawRequest(srv.url, {
      method: 'PUT',
      path: '/bucket/key',
      headers: [['content-md5', 'rL0Y20xC+Fzt72VPzMSk2A==']],
      body: new TextEncoder().encode('da'),
      sign: true,
      credentials: creds,
      region: 'us-east-1'
    })
    const head = srv.head()
    assert.ok(head.includes('Authorization: AWS4-HMAC-SHA256 Credential=AK/'), head)
    const authLine = head.split('\r\n').find((l) => l.startsWith('Authorization:')).toLowerCase()
    assert.ok(authLine.includes('content-md5'), 'content-md5 must be in SignedHeaders')
    assert.ok(!authLine.includes('content-length'), 'content-length must not be signed')
    assert.ok(head.includes('X-Amz-Content-Sha256: '), 'payload hash header present')
    assert.ok(head.includes('Content-Length: 2\r\n'), 'default content-length sent')
  } finally {
    await srv.close()
  }
})

test('authorization override wins over the signed value', async () => {
  const srv = await rawServer(OK_RESPONSE)
  try {
    await rawRequest(srv.url, {
      method: 'GET',
      path: '/bucket',
      headers: [['authorization', '']],
      sign: true,
      credentials: creds,
      region: 'us-east-1'
    })
    const head = srv.head()
    assert.ok(head.includes('authorization: \r\n'), 'override must win')
    assert.ok(!head.includes('AWS4-HMAC-SHA256'), 'signed value must not leak')
  } finally {
    await srv.close()
  }
})

test('response parsing: content-length, chunked, read-to-close', () => {
  const cl = parseResponse(Buffer.from('HTTP/1.1 200 OK\r\nContent-Length: 5\r\n\r\nhelloEXTRA'))
  assert.equal(Buffer.from(cl.body).toString(), 'hello')

  const chunked = parseResponse(Buffer.from(
    'HTTP/1.1 200 OK\r\nTransfer-Encoding: chunked\r\n\r\n5\r\nhello\r\n6\r\n world\r\n0\r\n\r\n'))
  assert.equal(Buffer.from(chunked.body).toString(), 'hello world')

  const close = parseResponse(Buffer.from('HTTP/1.1 404 Not Found\r\nX: y\r\n\r\neverything until close'))
  assert.equal(close.status, 404)
  assert.equal(Buffer.from(close.body).toString(), 'everything until close')
})

test('parseXmlError', () => {
  const [code, msg] = parseXmlError(new TextEncoder().encode(
    '<?xml version="1.0"?><Error><Code>NoSuchKey</Code><Message>it&apos;s gone &amp; lost</Message></Error>'))
  assert.equal(code, 'NoSuchKey')
  assert.equal(msg, "it's gone & lost")
  assert.deepEqual(parseXmlError(new TextEncoder().encode('not xml')), ['', ''])
  assert.deepEqual(parseXmlError(new Uint8Array(0)), ['', ''])
})
